### ISSUE TITLE
bpf userspace process taking long time to exit with too many attach uprobes( 1000 to 2000)

### ISSUE THREAD

**Issue by nalreddy** [libbpf/libbpf#870]
Applcation program :
     Simple test application with functions names as op_1(), op_2()...... op_2500();    2500 op_functions
     Build gcc probe_limits.c -o probe_limits

Bfptracer :   attaches uprobes to given function or function pattern  limit at 2000, not even running application process. just attach uprobes and exits.

   build 
         make;g++ -g bpftracer.cpp -o bpftracer -lbpf -lelf
   
    How to run 
          ./bpftracer <absolutepathto appp> <functionpattern>  <numbe of uprobes to attach>
          
     note :  increase ulimit appropriately for sudo user.

      sudo   ./bpftrace  probe_limits   op_*  1000

Attached files 

**Issue :   with or without bpf object  teardown  in userspace code process taking long time to exit. Looks like taking long time to cleanup bpf_link detach. Mean while bpf userspace process becoming defunct . 
But eventually exits after couple of seconds to minutes depending on number of uprobes attached.**

Question : I am doing something wrong in bpf setup and tear down of skeleton code.
      or is it expected to take long time to cleanup.

Tried strace ,gdb code couldn't collect ,as process was hung and defunct.

**FYI**
ps uax |grep bpftracer
user1   1102286 11.1  0.0      0     0 pts/2    D+   04:36   0:02 [bpftracer]

bpftool map show/list  , shows empty after process becoming defunct

cat /proc/1102286/status
Name:   bpftracer
State:  D (disk sleep)
Tgid:   1102286

**System details** 
 uname  -a
Linux  host1 6.5.0 #2 SMP PREEMPT_DYNAMIC Mon Oct  9 10:54:02 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
gcc version 9.5.0 (Ubuntu 9.5.0-3ubuntu1)
libbpf.so.1.4.0

Please do 
[libbpf_cleanup_issue.zip](https://github.com/user-attachments/files/17929767/libbpf_cleanup_issue.zip)
let me know how to debug further or need more details to help me out.

**Comment by nalreddy** [comment]
attached application and bpftracer files again 

build
make;g++ -g bpftracer.cpp -o bpftracer -lbpf -lelf

How to run 
      ./bpftracer <absolutepathto appp> <functionpattern>  <numbe of uprobes to attach>
      
 note :  increase ulimit appropriately for sudo user.

  sudo   ./bpftrace  probe_limits   op_*  1000

[libbpf_cleanup_issue.zip](https://github.com/user-attachments/files/17929781/libbpf_cleanup_issue.zip)

**Comment by anakryiko** [comment]
Yes, there is a bunch of work that's being done by the kernel to detach uprobes. This is not libbpf slowness, but rather kernel. There were a lot of changes in uprobes subsystem recently, so with newer kernels this hopefully is a bit slower, but I'm not sure it will be completely resolved.
